src/rpc.rs

```diff
@@ -0,0 +1,364 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::thread::{self, Builder, JoinHandle};
+
+pub const RPC_PORT: u16 = 8899;
+
+pub struct JsonRpcService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl JsonRpcService {
+    pub fn new(bank: Arc<Bank>, rpc_addr: SocketAddr) -> Self {
+        let request_processor = JsonRpcRequestProcessor::new(bank);
+        let thread_hdl = Builder::new()
+            .name("solana-jsonrpc".to_string())
+            .spawn(move || {
+                let mut io = MetaIoHandler::default();
+                let rpc = RpcSolImpl;
+                io.extend_with(rpc.to_delegate());
+
+                let server = ServerBuilder::new(io)
+                    .meta_extractor(move |_req: &hyper::Request| Meta {
+                        request_processor: Some(request_processor.clone()),
+                    })
+                    .threads(4)
+                    .cors(DomainsValidation::AllowOnly(vec![
+                        AccessControlAllowOrigin::Any,
+                    ]))
+                    .start_http(&rpc_addr)
+                    .unwrap();
+                server.wait();
+                ()
+            })
+            .unwrap();
+        JsonRpcService { thread_hdl }
+    }
+}
+
+impl Service for JsonRpcService {
+    fn thread_hdls(self) -> Vec<JoinHandle<()>> {
+        vec![self.thread_hdl]
+    }
+
+    fn join(self) -> thread::Result<()> {
+        self.thread_hdl.join()
+    }
+}
+
+#[derive(Clone)]
+pub struct Meta {
+    pub request_processor: Option<JsonRpcRequestProcessor>,
+}
+impl Metadata for Meta {}
+impl Default for Meta {
+    fn default() -> Self {
+        Meta {
+            request_processor: None,
+        }
+    }
+}
+
+build_rpc_trait! {
+    pub trait RpcSol {
+        type Metadata;
+
+        #[rpc(meta, name = "solana_confirmTransaction")]
+        fn confirm_transaction(&self, Self::Metadata, String) -> Result<bool>;
+
+        #[rpc(meta, name = "solana_getBalance")]
+        fn get_balance(&self, Self::Metadata, String) -> Result<(String, i64)>;
+
+        #[rpc(meta, name = "solana_getFinality")]
+        fn get_finality(&self, Self::Metadata) -> Result<usize>;
+
+        #[rpc(meta, name = "solana_getLastId")]
+        fn get_last_id(&self, Self::Metadata) -> Result<String>;
+
+        #[rpc(meta, name = "solana_getTransactionCount")]
+        fn get_transaction_count(&self, Self::Metadata) -> Result<u64>;
+
+        // #[rpc(meta, name = "solana_sendTransaction")]
+        // fn send_transaction(&self, Self::Metadata, String, i64) -> Result<String>;
+    }
+}
+
+pub struct RpcSolImpl;
+impl RpcSol for RpcSolImpl {
+    type Metadata = Meta;
+
+    fn confirm_transaction(&self, meta: Self::Metadata, id: String) -> Result<bool> {
+        let signature_vec = bs58::decode(id)
+            .into_vec()
+            .expect("base58-encoded public key");
+
+        if signature_vec.len() != mem::size_of::<Signature>() {
+            Err(Error::invalid_request())
```

Comment:
Use early returns to handle edge cases, so that the main case don't need to be nested under `else`. The value-add here is more obvious when there are multiple edge cases, but we should do it here too to keep the codebase consistent.